flink-cdc-connect/flink-cdc-source-connectors/flink-connector-mysql-cdc/src/main/java/com/ververica/cdc/connectors/mysql/debezium/DebeziumUtils.java

```diff
@@ -233,4 +236,75 @@ private static Map<String, String> querySystemVariables(
 
         return variables;
     }
+
+    public static BinlogOffset findBinlogOffset(long targetMs, MySqlConnection connection) {
+        MySqlConnection.MySqlConnectionConfiguration config = connection.connectionConfig();
+        BinaryLogClient client =
+                new BinaryLogClient(
+                        config.hostname(), config.port(), config.username(), config.password());
+
+        List<String> binlogFiles = new ArrayList<>();
+        JdbcConnection.ResultSetConsumer rsc =
+                rs -> {
+                    while (rs.next()) {
+                        String fileName = rs.getString(1);
+                        long fileSize = rs.getLong(2);
+                        if (fileSize > 0) {
+                            binlogFiles.add(fileName);
+                        }
+                    }
+                };
+
+        try {
+            connection.query("SHOW BINARY LOGS", rsc);
+            LOG.info("Total search binlog: {}", binlogFiles);
+
+            if (binlogFiles.isEmpty()) {
+                return BinlogOffset.ofBinlogFilePosition("", 0);
+            }
+
+            int index = searchBinlogOffset(client, targetMs, binlogFiles);
+            String binlogName = binlogFiles.get(index);
+            return BinlogOffset.ofBinlogFilePosition(binlogName, 0);
+        } catch (Exception e) {
+            throw new FlinkRuntimeException(e);
+        }
+    }
+
+    public static int searchBinlogOffset(
+            BinaryLogClient client, long targetMs, List<String> binlogFiles) throws IOException {
+        int startIdx = 0;
+        int endIdx = binlogFiles.size() - 1;
+
+        while (startIdx <= endIdx) {
+            int mid = startIdx + (endIdx - startIdx) / 2;
+            FindOffsetListener midBinlog = getBinlogOffsetListener(client, binlogFiles.get(mid));
+            long midTs = midBinlog.getFirstEventTs();
+            if (midTs < targetMs) {
+                startIdx = mid + 1;
+            } else if (targetMs < midTs) {
+                endIdx = mid - 1;
+            } else {
+                return mid;
+            }
+        }
+
+        return endIdx;
+    }
+
+    public static FindOffsetListener getBinlogOffsetListener(
+            BinaryLogClient client, String binlogFile) throws IOException {
```

Comment:
Still think it's better to return the timestamp directly as user who calls this method should not care about the listener or something else.

Besides, why not use an anonymous inner class here? The listener should be a kind of one-time utility just like the `JdbcConnection.ResultSetConsumer` above.

Reply:
Ok, I will fit.

Reply:
@whhe I am confused because seems an anonymous inner cannot return a result to outside directly, or we need use a collection outside like JdbcConnection.ResultSetConsumer to EventListener just for result a single timestamp?
`BinaryLogClient.EventListener eventListener = new BinaryLogClient.EventListener() {
            private long firstEventTs = 0L;
            @Override
            public void onEvent(Event event) {
                EventData data = event.getData();
                if (data instanceof RotateEventData) {
                    // We skip RotateEventData because it does not contain the timestamp we are interested
                    // in.
                    return;
                }

                EventHeaderV4 header = event.getHeader();
                long timestamp = header.getTimestamp();
                if (timestamp > 0) {
                    this.firstEventTs = timestamp;
                    try {
                        client.disconnect();
                    } catch (IOException e) {
                        throw new RuntimeException(e);
                    }
                }
            }

            public long getFirstEventTs() {           -- when use an anonymous inner class, we cannot call the method
                return this.firstEventTs;                  
            }
        };`

Reply:
Yes, we can add a list or something else here to receive the timestamp, and the demo written by GPT which uses `ArrayBlockingQueue<>(1)` looks good to me.

Reply:
@whhe Thanks for advice, ArrayBlockingQueue is ok to implement it, and I also give a new commit to use `private static class` and hidden the listener to outside，is it ok to you? or I will give the ArrayBlockingQueue commit later.

Reply:
> Yes, we can add a list or something else here to receive the timestamp, and the demo written by GPT which uses `ArrayBlockingQueue<>(1)` looks good to me.

I have done, pls review again : )

Reply:
Sorry for the late reply, a inner static class is also good here. I'll take a look on the latest commit soon.